### PR TITLE
fix manual componentname concatenation

### LINKED ISSUE #95: {componentName} interpolated just one time
I just started to write my config according to the [docs](https://github.com/postcss/postcss-bem-linter#componentselectors) regarding of interpolation of `{componentName}`.

But it just works for single interpolations, if I try to interpolate it multiple times, it throws:
`Error: "^\.{componentName}(?:__{componentName})*(?:--{componentName})?|\.(?:is|has)-{componentName}$" does not produce a valid regular expression`

**my config**
````json
{
  "rules": {
    "plugin/selector-bem-pattern": {
      "componentName": "[a-z0-9]+(?:-[a-z0-9]+)*",
      "componentSelectors": {
        "initial": "^\\.(?:no-)?js|\\.(?:is|has)-{componentName}|\\.{componentName}(?:__{componentName})*(?:--{componentName})?$",
        "combined": "^\\.{componentName}(?:__{componentName})*(?:--{componentName})?|\\.(?:is|has)-{componentName}$"
      }
    },
  }
}
````

**error**
````sh
/.../node/v5.10.1/bin/node /.../style-guide/node_modules/stylelint/dist/cli.js -f json --stdin-filename header.scss
Error: "^\.{componentName}(?:__{componentName})*(?:--{componentName})?|\.(?:is|has)-{componentName}$" does not produce a valid regular expression
    at /.../style-guide/node_modules/postcss-bem-linter/lib/to-interpolated-regexp.js:8:15
    at module.exports (/.../style-guide/node_modules/postcss-bem-linter/lib/validate-selectors.js:25:61)
    at checkRule (/.../style-guide/node_modules/postcss-bem-linter/index.js:78:7)
    at /.../style-guide/node_modules/postcss-bem-linter/index.js:45:9
    at Array.forEach (native)
    at /.../style-guide/node_modules/postcss-bem-linter/index.js:42:14
    at /.../style-guide/node_modules/postcss/lib/container.js:241:28
    at /.../style-guide/node_modules/postcss/lib/container.js:148:26
    at Root.each (/.../style-guide/node_modules/postcss/lib/container.js:114:22)
    at Root.walk (/.../style-guide/node_modules/postcss/lib/container.js:147:21)

Process finished with exit code 1
````

# Cause of issue

The Issue is caused by https://github.com/postcss/postcss-bem-linter/blob/master/lib/to-interpolated-regexp.js#L6, which should use `Array.join()` not manual concatenation.

### PR DESCRIPTION
**Pull request by AndyOGo**
fixes #95

### CODE REVIEW

**Comment by AndyOGo** [comment]
@davidtheclark 
Did you have any time to look into this?

**Comment by davidtheclark** [comment]
@AndyOGo: Yeah, within the next few days I will try to get to this.

**Comment by AndyOGo** [comment]
@davidtheclark 
Cool, thank you :)

File: lib/to-interpolated-regexp.js [hunk 1]
```diff
@@ -3,7 +3,7 @@ module.exports = function(source) {
     var splitSource = source.split('{componentName}');
     return function(componentName) {
       try {
-        return new RegExp(splitSource.length > 1 ? splitSource[0] + componentName + splitSource[1] : splitSource[0]);
+        return new RegExp(splitSource.join(componentName));
```
Comment (davidtheclark) [inline lib/to-interpolated-regexp.js]: I'm not sure this logic is right ... Can you please provide tests for this change, then we'll know whether it works correctly?

Comment (AndyOGo) [inline lib/to-interpolated-regexp.js]: Hmm, sure if you really need to see a failing test here, I can do it.

But to be honest, just check the docs of [`string.prototpe.split()`](String.prototype.split()) and [`Array.prototype.join()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join)

By splitting the `source` variable above, you get an Array of possible more than one items.
Hence you can just manually concatenate. Especially the check `splitSource.length > 1` here is very error prone.

Comment (davidtheclark) [inline lib/to-interpolated-regexp.js]: Yes, we do really need to have a test for this, whether it passes or fails.

Comment (AndyOGo) [inline lib/to-interpolated-regexp.js]: @davidtheclark 
I just pushed a test case which uses two custom configs, one which uses the `{componentName}` interpolation and one which doesn't.

**Comment by davidtheclark** [comment]
Thanks for adding the tests. Turning over to @simonsmith.

File: test/fixtures/bem-hyphenated.css [hunk 2]
```diff
@@ -0,0 +1,19 @@
+.block-name {}
```
Comment (simonsmith) [inline test/fixtures/bem-hyphenated.css]: Are we missing a `/** @define block-name */` here?

**Comment by AndyOGo** [comment]
@davidtheclark @simonsmith 

I just realized I misunderstood how the `{componentName}` interpolation works, as described here:
> A single string that provides a valid pattern for the RegExp() constructor when {componentName} is interpolated with the defined component's name, e.g.

I thought it will interpolated the RegEx given for `componentName`, but it does interpolate the vale retrived from the RegEx match. Which obviously can't work.

Despite of my misunderstanding the above is still a bug.

May I suggest an API change, so that not the RegEx match is interpolated, instead the regEx definition?

**Comment by davidtheclark** [comment]
@AndyOGo Please file a separate issue for your `{componentRegex}` interpolation proposal — which seems like a fine idea to me.

**Comment by AndyOGo** [comment]
@davidtheclark 
Thanks for your reply. I will close this and file a separate one.

**Pull request closed without merging**